File: PRNumberBiz/app/src/main/java/com/pplus/prnumberbiz/core/sns/google/GoogleUtil.java
package com.pplus.prnumberbiz.core.sns.google;

import android.app.Activity;
import android.content.Intent;
import androidx.annotation.NonNull;
import androidx.fragment.app.FragmentActivity;
import android.widget.Toast;
import com.google.android.gms.auth.api.Auth;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.auth.api.signin.GoogleSignInResult;
import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.api.GoogleApiClient;
import com.pplus.prnumberbiz.R;

/**
 * Created by 김종경 on 2016-06-30.
 */

public class GoogleUtil {

    private static final int RC_SIGN_IN = 9000;

    /**
     * GoogleSignInResult 리스너
     */
    public interface GoogleSignListener{

        public void handleSignInResult(GoogleSignInResult result);
    }

    private static GoogleApiClient mGoogleApiClient;

    private static GoogleSignListener mListener;

    private static void validateServerClientID(Activity activity){

        String serverClientId = activity.getString(R.string.default_web_client_id);
        String suffix = ".apps.googleusercontent.com";
        if(!serverClientId.trim().endsWith(suffix)) {
            String message = "Invalid server client ID in strings.xml, must end with " + suffix;
            Toast.makeText(activity, message, Toast.LENGTH_LONG).show();
        }
    }

    /**
     * 초기화, listener등록 후 sign in 시도
     **/
    public static void init(FragmentActivity activity, GoogleSignListener listener){

        mListener = listener;

        validateServerClientID(activity);

        if(mGoogleApiClient == null || !mGoogleApiClient.isConnected()){
            GoogleSignInOptions gso = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN).requestIdToken(activity.getString(R.string.default_web_client_id)).requestServerAuthCode(activity.getString(R.string.default_web_client_id), false).requestEmail().build();
            mGoogleApiClient = new GoogleApiClient.Builder(activity).enableAutoManage(activity /* FragmentActivity */, new GoogleApiClient.OnConnectionFailedListener(){

                @Override
                public void onConnectionFailed(@NonNull ConnectionResult connectionResult){

                }
            } ).addApi(Auth.GOOGLE_SIGN_IN_API, gso).build();
        }

        Intent signInIntent = Auth.GoogleSignInApi.getSignInIntent(mGoogleApiClient);
        activity.startActivityForResult(signInIntent, RC_SIGN_IN);
    }

    public static void signOut(){
        if(mGoogleApiClient.isConnected()){
            Auth.GoogleSignInApi.signOut(mGoogleApiClient);
        }
    }

    /**
     * onActivityResult 함수에서 호출
     **/
    public static void onActivityResult(int requestCode, int resultCode, Intent data){

        switch (requestCode) {
            case RC_SIGN_IN:
                GoogleSignInResult result = Auth.GoogleSignInApi.getSignInResultFromIntent(data);
                if(mListener != null) {
                    mListener.handleSignInResult(result);
                }
                break;
        }
    }
}

File: PRNumber/app/src/main/java/com/pplus/prnumberuser/core/sns/google/GoogleUtil.kt
package com.pplus.prnumberuser.core.sns.google

import android.app.Activity
import android.content.Intent
import android.widget.Toast
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.fragment.app.FragmentActivity
import com.google.android.gms.auth.api.signin.*
import com.google.android.gms.common.api.ApiException
import com.pplus.prnumberuser.R

/**
 * Created by 김종경 on 2016-06-30.
 */
object GoogleUtil {
    private const val RC_SIGN_IN = 9000
    private var mGoogleSignInClient: GoogleSignInClient? = null
    var mListener : GoogleSignListener? = null
    private fun validateServerClientID(activity: Activity) {
        val serverClientId = activity.getString(R.string.default_web_client_id)
        val suffix = ".apps.googleusercontent.com"
        if (!serverClientId.trim { it <= ' ' }.endsWith(suffix)) {
            val message = "Invalid server client ID in strings.xml, must end with $suffix"
            Toast.makeText(activity, message, Toast.LENGTH_LONG).show()
        }
    }

    /**
     * 초기화, listener등록 후 sign in 시도
     */
    fun init(activity: FragmentActivity, listener: GoogleSignListener?) {
        mListener = listener
        validateServerClientID(activity)

        val gso = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
            .requestIdToken(activity.getString(R.string.default_web_client_id))
            .requestServerAuthCode(activity.getString(R.string.default_web_client_id), false)
            .requestEmail().build()

        mGoogleSignInClient = GoogleSignIn.getClient(activity, gso)

    }

    fun launch(launcher: ActivityResultLauncher<Intent>){
        val signInIntent = mGoogleSignInClient!!.signInIntent
        launcher.launch(signInIntent)
    }

    fun signOut() {
        mGoogleSignInClient?.signOut()
    }


    /**
     * GoogleSignInResult 리스너
     */
    interface GoogleSignListener {
        fun handleSignInResult(account: GoogleSignInAccount?)
    }
}